File: mortgage_calculator.js
var PROPERTY_TAX_RATE = 0.015;
var home_value = 1000000.0;
var deposit = 500000;
var yearly_property_tax = home_value * PROPERTY_TAX_RATE;
var monthly_property_tax = yearly_property_tax / 12.0;
var monthly_property_tax_adjustment = monthly_property_tax * 0.36;
var principal = 500000.0;
var monthly_interest_rate = 0.03 / 12;
var total_payments = 360;
var num = principal * (monthly_interest_rate * Math.pow( 1 + monthly_interest_rate, total_payments));
var denom = Math.pow(1 + monthly_interest_rate, total_payments) - 1;
var monthly_mortgage_payment = num / denom;
var net_cost = monthly_mortgage_payment + monthly_property_tax;
console.log( 'monthly_mortgage_payment', monthly_mortgage_payment );
console.log( 'mortgage + property tax', net_cost );
console.log( "savings from property tax writeoff", monthly_property_tax_adjustment);
console.log( "adjusted cost", net_cost - monthly_property_tax_adjustment );
console.log( "monthly mortgage payment", monthly_mortgage_payment);

function calculateMonthlyPayments(target_monthly_payment, hoa) {
  var payments = [];
  running_principal = principal
  running_value = home_value

  for (var month = 0; month < 360; month++) {
    var amounts = {};
    amounts['month'] = month;
    amounts['interest'] = monthly_interest_rate * running_principal
    amounts['principal'] = monthly_mortgage_payment - amounts['interest']
    amounts['principal'] += (target_monthly_payment - monthly_mortgage_payment - hoa)
    running_principal -= amounts['principal']
    amounts['property_tax'] = monthly_property_tax
    amounts['hoa'] = hoa
    amounts['deductible'] = amounts['interest'] + monthly_property_tax
    amounts['deduction_savings'] = amounts['deductible'] * 0.36
    amounts['adjusted_monthly_payment'] = target_monthly_payment - amounts['deduction_savings']
    amounts['running_principal'] = running_principal
    amounts['running_value'] = (running_value *= 1.00583) // 1.07 / 12 months
    amounts['ownership'] = home_value - running_principal
    amounts['ownership_percent'] = amounts['ownership'] / home_value
    amounts['total_paid'] = month * target_monthly_payment + deposit
    // profit_rate = amounts['running_value'] / home_value
    // this still doesn't seem right..
    amounts['running_profit'] = (running_value - amounts['total_paid']) * amounts['ownership_percent']
    payments.push(amounts);

    if (month == 7 * 12) {
      console.log(" ** 7 years, $"+target_monthly_payment+"/month, "+amounts['running_value']+", HOA: "+hoa);
      console.log(amounts);
    }
    months_to_payoff_by_monthly_payment[Math.round(target_monthly_payment)] = month;
    if (running_principal < 0) { break; }
  }

  return payments;
};

hoas = [0, 300, 500, 800];
target_monthly_payments = [monthly_mortgage_payment + monthly_property_tax, 4000, 5000];

months_to_payoff_by_monthly_payment = {}
for (var i = 0; i < hoas.length; i++) {
  var hoa = hoas[i];

  for (var j = 0; j < target_monthly_payments.length; j++) {
    var target_monthly_payment = target_monthly_payments[j];
    var payments = calculateMonthlyPayments(target_monthly_payment, hoa);
    console.log('payments', payments);
  }
  console.log(months_to_payoff_by_monthly_payment);
}

File: mortgage_calculator.rb
require 'byebug'
require 'csv'
require 'set'
require 'awesome_print'
require 'active_support/core_ext/integer/time' # for Numeric#months

class Mortgage
  TOTAL_PAYMENTS = 360
  HOME_VALUE_YEARLY_GROWTH_RATE = 0.04

  def initialize(home_value: , deposit: , hoa: , property_tax_rate: , target_monthly_payment: , first_month: )
    @home_value = home_value
    @deposit = deposit
    @hoa = hoa
    @property_tax_rate = property_tax_rate
    @monthly_property_tax = property_tax_rate / 12.0
    @target_monthly_payment = target_monthly_payment
    @first_month = first_month
  end

  def monthly_interest_rate(month)
    fail NotImplementedError
  end

  def monthly_mortgage_payment(month)
    rate = monthly_interest_rate(month)
    (@home_value - @deposit) * (rate * ( 1 + rate) ** TOTAL_PAYMENTS) / ((1 + rate) ** TOTAL_PAYMENTS - 1)
  end

  def miniumum_monthly_payment(month)
    monthly_mortgage_payment(month) + @monthly_property_tax + @hoa
  end

  def perform
    deductions = Hash.new(0)
    running_principal = @home_value - @deposit
    running_value = @home_value
    running_interest = 0
    running_payments = 0
    monthly_payments = Hash.new(0)

    TOTAL_PAYMENTS.times do |month|
      date = @first_month + month.months
      minimum_payment = miniumum_monthly_payment(month)
      monthly_payment = [minimum_payment, @target_monthly_payment].max
      monthly_payments[monthly_payment.to_i] += 1

      amounts = {}
      amounts[:month] = month
      amounts[:monthly_payment] = monthly_payment
      running_payments += monthly_payment

      amounts[:interest_rate] = monthly_interest_rate(month) * 12
      amounts[:interest] = monthly_interest_rate(month) * running_principal
      running_interest += amounts[:interest]
      amounts[:running_interest] = running_interest

      amounts[:principal] = monthly_payment - amounts[:interest] - @monthly_property_tax - @hoa
      running_principal -= amounts[:principal]
      amounts[:running_principal] = running_principal

      amounts[:deduction_savings] = (amounts[:interest] + @monthly_property_tax) * 0.36
      deductions[date.year] += amounts[:interest] + @monthly_property_tax
      amounts[:adjusted_monthly_payment] = monthly_payment - (amounts[:interest] + @monthly_property_tax) * 0.36

      running_value *= (1 + (HOME_VALUE_YEARLY_GROWTH_RATE / 12))
      amounts[:running_value] = running_value
      amounts[:ownership_percent] = 100.0 * (@home_value - running_principal) / @home_value

      if running_principal <= 0
        summary = {
          target_monthly_payment: @target_monthly_payment,
          loan: self.class.name,
          years: (month / 12.0).round(1),
          loan_amount: (@home_value - @deposit).round,
          interest: running_interest.round,
          payments: running_payments.round,
          paid: (@deposit + running_payments).round,
          value: running_value.round,
          percent_interest: (100.0 * running_interest / running_payments).round(1)
        }
        summary[:multiplier] = (summary[:payments] / summary[:loan_amount].to_f).round(2)
        summary[:projected_profit] = summary[:value] - summary[:paid]
        summary[:deductions] = Hash[deductions.map {|k,v|[k,v.to_i]}]
        summary[:monthly_payments] = monthly_payments

        # summary.except(:deductions, :monthly_payments).keys
        return summary
      end
    end
  end
end

class FiveFiveMortgage < Mortgage
  def monthly_interest_rate(month)
    if month / 12 < 5
      0.04199 / 12
    elsif month / 12 < 10
      0.06199 / 12
    elsif month / 12 < 15
      0.08199 / 12
    else # lifetime cap of 5% over the initial rate
      0.10199 / 12
    end
  end
end

class AdjustableRateMortgage < Mortgage
  LIBOR = 0.0176
  MARGIN = 0.035

  def intro_rate
    fail NotImplementedError
  end

  def cap_rate
    self.intro_rate + 0.05
  end

  def monthly_interest_rate(month)
    if month / 12 < self.change_year
      intro_rate / 12
    else
      #(LIBOR + MARGIN) / 12
      self.cap_rate / 12
    end
  end
end

class ThreeOneMortgage < AdjustableRateMortgage
  def intro_rate
    0.04044
  end

  def change_year
    3
  end
end

class FiveOneMortgage < AdjustableRateMortgage
  def intro_rate
    0.04151
  end

  def change_year
    5
  end
end

class SevenOneMortgage < AdjustableRateMortgage
  def intro_rate
    0.04275
  end

  def change_year
    7
  end
end

class ThirtyYearMortgage < Mortgage
  def monthly_interest_rate(_)
    0.0425 / 12
  end
end

class MortageComparer
  CLASSES = [ThirtyYearMortgage, FiveFiveMortgage, SevenOneMortgage]
  def self.perform
    summary = {}
    CLASSES.each do |klass|
      klass_key = klass.to_s.split(' ').first
      (4000..10000).step(500).each do |target_monthly_payment|
        params = {
          home_value: 1_250_000,
          deposit: 641_000,
          hoa: 350,
          property_tax_rate: 0.01179,
          target_monthly_payment: target_monthly_payment,
          first_month: Date.new(2017, 5, 1)
        }
        summary[klass_key] ||= {}
        summary[klass_key][target_monthly_payment] = klass.new(params).perform
      end
    end
    summary
  end
end

ap MortageComparer.perform
